rust/lance-encoding/src/encodings/physical/bitpack.rs

```diff
@@ -0,0 +1,220 @@
+                bytes_to_skip -= buf_len;
+                continue;
+            }
+
+            // the index of the current byte being read from the source buffer
+            let mut src_idx = bytes_to_skip as usize;
+
+            // bit offset within the current source byte
+            let mut src_offset = 0;
+
+            while src_idx < src.len() && rows_taken < num_rows {
+                rows_taken += 1;
+                let mut curr_mask = mask; // copy mask
+                
+                // current source byte being written to destination
+                let mut curr_src = src[src_idx] >> src_offset & curr_mask as u8;
+                
+                // how many bits from the current source value have been written to destination
+                let mut src_bits_written = 0;
+
+                // the offset within the current destination byte to write to
+                let mut dst_offset = 0;
+
+                while src_bits_written < self.bits_per_value {
+                    // write bits from current source byte into destination
+                    dst[dst_idx] += (curr_src) << dst_offset;
+                    let bits_written = (self.bits_per_value - src_bits_written)
+                        .min(8 - src_offset)
+                        .min(8 - dst_offset);
+                    src_bits_written += bits_written;
+                    dst_offset += bits_written;
+                    src_offset += bits_written;
+
+                    if dst_offset == 8 {
+                        dst_idx += 1;
+                        dst_offset = 0;
+                    }
+
+                    if src_offset == 8 {
+                        src_idx += 1;
+                        src_offset = 0;
+                        curr_mask >>= 8;
+                        if src_idx == src.len() {
+                            break;
+                        }
+                        curr_src = src[src_idx] & curr_mask as u8;
+                    }
+                }
+
+                // advance destination offset to the next location
+                // note that we don't need to do this if we wrote the full number of bits
+                // because source index would have been advanced by the inner loop above
+                if self.uncompressed_bits_per_value != self.bits_per_value {
+                    let mut partial_bytes_written = self.bits_per_value / 8;
+
+                    // if we didn't write the full byte for the last byte, increment by one because
+                    // we wrote a partial byte
+                    if self.uncompressed_bits_per_value % self.bits_per_value != 0 {
+                        partial_bytes_written += 1;
+                    }
+                    dst_idx += (byte_len as u64 - partial_bytes_written + 1) as usize;
+                }
+            }
+        }
+    }
+
+    fn num_buffers(&self) -> u32 {
+        // TODO ask weston what this is about
+        1
+    }
```

Comment:
1 is correct.  There are some cases (e.g. dictionary encoding) where we encode 1 input buffer into 2 output buffers.